### DISCUSSION TITLE
How to make dual (BIOS & UEFI) USB or ISO

### DISCUSSION THREAD

**Discussion by jeffshead** [ipxe/ipxe#1182] [Q&A]
I'm trying to make a single USB or ISO image that can be used to support both EFI and legacy BIOS computers.

The following commands work to build the image but the keyboard stops working if I use the image and boot efi:

```
# for systems with BIOS
make bin/acm--ncm--ecm--axge.usb ... # your options may come here

# copy MBR and boot loader image on /dev/sdb disk beginning.
dd if=bin/acm--ncm--ecm--axge.usb of=/dev/sdb oflag=sync

# for systems with UEFI
make bin-x86_64-efi/acm--ncm--ecm--axge.efi ... # your options may come here

# create ESP partition, FAT32 filesystem and copy boot loader so UEFI firmware can find it
# I left 5Mb offset as MBR and /dev/sdb4 partition occupies 4095 sectors from beginning of disk (~2MB)
# partition size - choose any fits your needs. In my case ending point is 311M
parted -s /dev/sdb mkpart primary fat32 10MiB 311MiB
parted -s /dev/sdb set 1 esp on
partprobe /dev/sdb

mkfs.fat -F32 /dev/sdb1

mount /dev/sdb1 /mnt
mkdir -p /mnt/efi/boot/
cp bin-x86_64-efi/acm--ncm--ecm--axge.efi /mnt/efi/boot/bootx64.efi
umount /mnt
```
The above code came from this [post](https://gist.github.com/AdrianKoshka/5b6f8b6803092d8b108cda2f8034539a?permalink_comment_id=4606457#gistcomment-4606457).

According to this [post](https://github.com/ipxe/ipxe/discussions/308#discussioncomment-595628), uncommenting `//#define	USB_HCD_USBIO	/* Very slow EFI USB host controller */` in _/src/config/usb.h_ should resolve the issue but I get the following error when I try to make the efi file:

```
make bin-x86_64-efi/acm--ncm--ecm--axge.efi

ld: bin-x86_64-efi/blib.a(usbio.o):/home/jeff/ipxe/src/drivers/usb/usbio.c:1725: multiple definition of `usbio_driver'; bin-x86_64-efi/blib.a(efi_usb.o):/home/jeff/ipxe/src/interface/efi/efi_usb.c:1360: first defined here
make: *** [Makefile.housekeeping:1233: bin-x86_64-efi/acm--ncm--ecm--axge.efi.tmp] Error 1
rm bin-x86_64-efi/version.acm--ncm--ecm--axge.efi.o
```
What am I doing wrong?

**Comment by NiKiZe** [comment]
To generate dual boot images use `util/genfsimg` as described in https://github.com/ipxe/ipxe/discussions/1155#discussioncomment-8480915

What you seem to be trying to generate is a USB NIC only boot image? To get access to USB interfaces, iPXE has its own USB drivers, which does disconnect devices from BIOS/firmware drivers, iPXE should however have its own USB keyboard driver.

Are you using latest git master? Do you have any other local modifications?

What is your end goal, which NIC are you trying to boot from? (USB Vendor and device id or PCIID)

**Comment by jeffshead** [comment]
>USB NIC only

Yes, I'm trying to PXE boot some computers with USB NICs.

>Are you using latest git master?

Yes

>Do you have any other local modifications?

No

>What is your end goal, which NIC are you trying to boot from?

Plugable, Dell, Kanex, Acer brands. Most have the Realtek RTL8153 chipset. The others have the Asix AX88179.

The computers that I'm trying to PXE boot do not have built-in Ethernet NICs. Their firmwares do not support PXE booting with any of these USB NICs. As a workaround, I'm trying to create a bootable iPXE USB thumb drive. As I stated previously, I successfully created an iPXE USB thumb drive that supports both BIOS and EFI but once booted, the keyboard no longer works for EFI PXE boots. I don't encounter this issue with BIOS PXE boots. Because of this issue, I tried uncommenting that line in the _usb.h_ file as I described previously, but doing so causes errors during the build.

The USB thumb drive doesn't have to be for USB NICs only, but it does need to support USB NICs.

**Comment by mcb30** [comment]
> The computers that I'm trying to PXE boot do not have built-in Ethernet NICs. Their firmwares do not support PXE booting with any of these USB NICs. As a workaround, I'm trying to create a bootable iPXE USB thumb drive. As I stated previously, I successfully created an iPXE USB thumb drive that supports both BIOS and EFI but once booted, the keyboard no longer works for EFI PXE boots.

Please clarify "keyboard no longer works".  Do you mean that the keyboard no longer works within the iPXE shell (or menu), or that the keyboard is no longer working within some program that you have used iPXE to boot?

**Comment by jeffshead** [comment]
The keyboard stops working as soon as I get to the iPXE shell.

**Comment by mcb30** [comment]
> The keyboard stops working as soon as I get to the iPXE shell.

Thanks.

Could you please try building with `DEBUG=usb,usbhid,usbkbd`?  You can do this without the use of `parted`, `partprobe`, etc using:
```
make bin/acm--ncm--ecm--axge.lkrn bin-x86_64-efi/acm--ncm--ecm--axge.efi DEBUG=usb,usbhid,usbkbd

./util/genfsimg -o boot.iso bin/acm--ncm--ecm--axge.lkrn bin-x86_64-efi/acm--ncm--ecm--axge.efi
```
to generate the hybrid USB/ISO image `boot.iso`.

**Comment by jeffshead** [comment]
@mcb30 
I was able to use the commands that you provided to create the image but when I boot from it, there is only an option to boot from BIOS, not efi.
![PXL_20240403_010215879](https://github.com/ipxe/ipxe/assets/349651/75637d97-2e37-4d03-a8b5-0aa9cff3cf10)

Also, when the USB drive is attached to a Windows computer, it uses two drive letters and shows up as a USB drive and a DVD.

I tried twice and got the same results both times.
![2024-04-02_21-09-25](https://github.com/ipxe/ipxe/assets/349651/b792a055-64d3-41e0-8570-8d90d45bae7b)

**Reply by mcb30** [reply]
The image built with those commands will be a hybrid ISO/USB image (as you originally requested).  It's possible that your BIOS does not handle this well.  Try building just a USB image instead by changing `-o boot.iso` to `-o boot.img`.

**Comment by jeffshead** [comment]
Thank you for your time!

What's the best way to write the image to the USB drive? I don't know what I'm doing wrong. I've tried dd, isohybrid and Rufus. I get varying results but none (other than my original post) produce a USB thumb drive that shows the correct options under both the legacy and UEFI boot menus when using the .iso or .img file.

Using `-o boot.img` created a USB drive where the USB drive shows up as a **Diskette Drive** under the legacy menu:
![PXL_20240403_132855207 (Custom)](https://github.com/ipxe/ipxe/assets/349651/e98b04f7-6ec1-4153-965e-341697f3b2c6)
However, if I select **Diskette Drive**, it does perform the legacy boot and the keyboard does work.

Below are screen caps depicting the debug info when legacy booting (keyboard works):
![bios_debug_resized](https://github.com/ipxe/ipxe/assets/349651/a897cd90-fb8a-4ba5-93ce-b026a035f125)

The image below depicts the debug info when I do the UEFI boot:
![PXL_20240403_133037316](https://github.com/ipxe/ipxe/assets/349651/d7b17f29-1880-415f-9a7a-81c2f8d9aab4)

The UEFI boot is successful but the keyboard does not work.

**Comment by mcb30** [comment]
Thank you for the well described and relevant screenshots!

> What's the best way to write the image to the USB drive? I don't know what I'm doing wrong. I've tried dd, isohybrid and Rufus. I get varying results but none (other than my original post) produce a USB thumb drive that shows the correct options under both the legacy and UEFI boot menus when using the .iso or .img file.

Using `dd` is correct.

> Using `-o boot.img` created a USB drive where the USB drive shows up as a **Diskette Drive** under the legacy menu:

That makes some sense: `genfsimg` uses `mformat` to create a FAT filesystem within the unpartitioned image, and your BIOS is choosing to interpret any unpartitioned disk as "Diskette".

> The UEFI boot is successful but the keyboard does not work.

In the BIOS build you have `USB_KEYBOARD` still enabled.  In the UEFI build I think you have a local customisation to disable `USB_KEYBOARD` since the screenshot shows `using driver usbio` rather than `using driver kbd`, i.e. the USB `kbd` driver is not present and the device is therefore being offered for consumption by existing UEFI drivers via the `usbio` interface.

Unfortunately the UEFI standard's EFI_USB_IO_PROTOCOL is so badly designed that several OEMs chose to use an alternative, completely proprietary, and undocumented interface instead.  You can't really blame them, given the design flaws: https://github.com/ipxe/ipxe/blob/59f27d69358efc919b50760f3d6dac0b637b5488/src/drivers/usb/usbio.c#L43-L49

This has the side effect that the OEM's USB drivers (such as the keyboard driver) are unable to bind to the standard EFI_USB_IO_PROTOCOL interfaces exposed by iPXE.  It's therefore not that surprising that your keyboard doesn't work in this setup.

Could you please retry your UEFI test case **without** disabling `USB_KEYBOARD` in `config/usb.h`?

**Comment by jeffshead** [comment]
>I think you have a local customisation to disable USB_KEYBOARD since the screenshot shows using driver usbio rather than using driver kbd

I havn't customized any files nor have I added any files. For this latest test, I redownloaded the source files for a clean start. 

>Could you please retry your UEFI test case without disabling USB_KEYBOARD in config/usb.h?

That line is commented out by default. I did as you suggested and uncommented that line in config/usb.h before running any other commands.

I get the same results. Keybord still doesn't work. Below is a screenshot of UEFI boot:
![keyboard](https://github.com/ipxe/ipxe/assets/349651/ecbd1af4-62ee-4d73-b4a9-3d4a42c889b2)

**Comment by mcb30** [comment]
> > I think you have a local customisation to disable USB_KEYBOARD since the screenshot shows using driver usbio rather than using driver kbd
> 
> I havn't customized any files nor have I added any files. For this latest test, I redownloaded the source files for a clean start.

My mistake, sorry.  I lost track of what our current default UEFI configuration is.

For BIOS, we include `USB_KEYBOARD` support by default (set in `config/defaults/pcbios.h`).  For UEFI, we currently assume that the OEM's keyboard driver will continue to function with the EFI_USB_IO_PROTOCOL interface that we expose.

Could you try putting an explicit
```c
#define USB_KEYBOARD
```
in `config/usb.h` and see if that then works?

**Comment by jeffshead** [comment]
>in config/usb.h and see if that then works?

Keyboard still doesn't work. Below is screencap with explicit:

![define](https://github.com/ipxe/ipxe/assets/349651/7258aea5-a8a1-4a3d-bcbd-56a6b35577cb)

**Reply by NiKiZe** [reply]
You have 2 different versions of iPXE here, please test one at a time for working keyboard.

**Reply by jeffshead** [reply]
>You have 2 different versions of iPXE here, please test one at a time for working keyboard.

I'm sorry but I'm not understanding your comment. I do see **iPXE 1.21.1+** and **iPXE 1.0.0+** in my screenshots. Is one referring to the USB drive and the other to the server? I don't know how to make them match.

**Reply by mcb30** [reply]
@NiKiZe Good catch - I hadn't looked closely enough to notice that there were two iPXE binaries involved!

@jeffshead The output from `iPXE 1.21.1+...` is from the USB drive that you have just built.  In the latest screenshot, you can see the `using driver kbd`, as expected now that you have explicitly enabled `USB_KEYBOARD`.

The output from `iPXE 1.0.0+...` is from some other iPXE binary that your network is handing out as the file `iventoy_loader_16000`.  We have no idea what is in this binary.  Let's eliminate it from the test setup.

Please create a plain text file `shell.ipxe` containing just:
```
#!ipxe
shell
```
and add `EMBED=shell.ipxe` to your build command.  For example:
```
make bin/acm--ncm--ecm--axge.lkrn bin-x86_64-efi/acm--ncm--ecm--axge.efi DEBUG=usb,usbhid,usbkbd EMBED=shell.ipxe

./util/genfsimg -o boot.iso bin/acm--ncm--ecm--axge.lkrn bin-x86_64-efi/acm--ncm--ecm--axge.efi
```

That should give you a version of iPXE with an embedded script that will immediately drop to the iPXE shell (without displaying a "Press Ctrl-B..." prompt).  This will prevent your USB drive's version of iPXE from attempting to boot from the network, and so we won't end up loading your other unknown version of iPXE.  Let's test that first, so that we at least know which code we're testing.

**Reply by jeffshead** [reply]
I tested with and without adding `#define USB_KEYBOARD` to the **config/usb.h** file. The keyboard works either way, up to this point. I can type in commands. If I enter `autoboot`, the network version boots but of course the keyboard no longer works.

The screen cap below is without `#define USB_KEYBOARD` and without any other edits to **config/usb.h**.

![unedited](https://github.com/ipxe/ipxe/assets/349651/89b41bd4-3718-4164-99dd-2e8d254059c2)

**Reply by NiKiZe** [reply]
This means that your USB build works as it should.
As mentioned, your iventoy_loader_16000 is probably the one that is causing issues, you should probably not need to chain into that iPXE build at all?

Next step here is probably to look into which script it is running, and call that directly from your new build.

**Reply by jeffshead** [reply]
>Next step here is probably to look into which script it is running, and call that directly from your new build.

Unfortunately, it looks like iVentoy is closed source so I guess this is as far as we can go. I'll see if I can figure out how to move away from iVentoy. I started with it because it made PXE booting so simple to set up and it allowed me to use my current Windows DNS server with a single URL to boot both legacy and UEFI. 

Thank you @mcb30 and @NiKiZe for your time.

**Reply by NiKiZe** [reply]
They are using iPXE, the license does not allow closed source. It does allow not to publish sources if it is totally unmodified version of the code found here. But if that is the case you will be able to create a new version of iPXE that matches.

The question then become, is any embedded script part of the source or not.

**Reply by jeffshead** [reply]
On their [Github page](https://github.com/ventoy/PXE/tree/master/iPXE), they list the open source portion which has a folder (10 mo old) named **ipxe-bd13697**. I keyword searched `version` but I wasn't able to determine the iPXE version. I will attempt to make a USB drive using the version on their Github page.

**Reply by jeffshead** [reply]
> I will attempt to make a USB drive using the version on their Github page

Failed using same make commands. Bunch of errors. Mostly _"...architecture is not supported"_,  `error: operator '==' has no left operand` errors.

[terminal_output.txt](https://github.com/ipxe/ipxe/files/14946200/terminal_output.txt)

**Reply by NiKiZe** [reply]
> On their [Github page](https://github.com/ventoy/PXE/tree/master/iPXE), they list the open source portion which has a folder (10 mo old) named **ipxe-bd13697**. I keyword searched `version` but I wasn't able to determine the iPXE version. I will attempt to make a USB drive using the version on their Github page.

Let's assume it's based on
https://github.com/ipxe/ipxe/commit/bd13697

They do have a build.sh script, but nothing embedded there.

**Comment by jeffshead** [comment]
I used [bd13697](https://github.com/ipxe/ipxe/commit/bd13697) to create the USB thumb drive with `EMBED=shell.ipxe` and with no other modifications. The keyboard fails to work when booting UEFI.
![bd13697](https://github.com/ipxe/ipxe/assets/349651/cf0a24d9-c84e-48c0-b46e-975a2e1d89da)

Below is the unedited **config/usb.h**:
```
#include <config/defaults.h>

/*
 * USB host controllers (all enabled by default)
 *
 */
//#undef        USB_HCD_XHCI    /* xHCI USB host controller */
//#undef        USB_HCD_EHCI    /* EHCI USB host controller */
//#undef        USB_HCD_UHCI    /* UHCI USB host controller */
//#define       USB_HCD_USBIO   /* Very slow EFI USB host controller */

/*
 * USB peripherals
 *
 */
//#undef        USB_KEYBOARD    /* USB keyboards */
//#undef        USB_BLOCK       /* USB block devices */

/*
 * USB external interfaces
```

Next, I re-created the USB drive with `EMBED=shell.ipxe` and with the `#undef        USB_KEYBOARD    /* USB keyboards */` line uncommented in **config/usb.h**. I got the same results.

Next, I added `#define USB_KEYBOARD` to **config/usb.h** and the keyboard works in the shell:
![bd13697_define](https://github.com/ipxe/ipxe/assets/349651/8795d077-760d-4c24-aaf4-1e8c6d8b0f1a)

If I enter `autoboot`, the network version boots but of course the keyboard no longer works.

**Reply by NiKiZe** [reply]
What I would do is use that commit, apply their tree, check for changes, and maybe commit them, then rebase on current master.

When you create the usb build you will automatically get everything enabled, so no need to make any local usb configuration changes.

**Comment by mcb30** [comment]
@jeffshead From reading your posts, I think you have:

* Commit https://github.com/ipxe/ipxe/commit/bd13697 works only if you add `#define USB_KEYBOARD`
* Current master works with or without `#define USB_KEYBOARD`
 
It would be interesting to know which commit fixed the problem.  If you are interested, you can try following the instructions at https://ipxe.org/howto/bisect to identify this specific commit.

(This won't do anything to change the binary included in iVentoy: it would be purely for curiosity.)
